### ISSUE TITLE
Less error prone tuple assignments

### ISSUE THREAD

**Issue by axic** [ethereum/solidity#3314]
(This has been also mentioned by the recent Solidity audit under SOL-007.)

Currently tuple assignments can skip leading, trailing and positional variable names.  I'd propose to make these rules strict and require the user to specify the required number of return values by appropriate number of variable names. If omitted either accept an empty name or require `_`.

e.g. `(,a,,,,b,,,,,c,,,,,) = crazyfunction()` or `(_,a,_,_,_,_b,_,_,_,_,_c,_,_,_,_,_) = crazyfunction()`

If such a construct is deemed unreadable by the user that should trigger a refactoring of the code to avoid returning countless return values, which probably will benefit the code quality.

(`_` makes sense because some languages (like Lua) already do that, but it also conflicts with the modifier body marker `_`)

**Comment by chriseth** [comment]
So you would not allow a "match anything" marker?

I don't think that `_` conflicts with the function body marker in the modifier, because that is only valid at statement level.

We should perhaps turn `_` into a keyword for 0.5.0

**Comment by chriseth** [comment]
Ok, I think we agreed to remove the "match any number" feature.

Because of that, this is a breaking change.

Proposal for implementation:

 - warn about "match any number" until 0.5.0 and turn into error from then on
 - warn about missing component until 0.5.0 and turn into error from then on
 - expressions of the form `(a,)` are still fine (they distinguish 1-element tuples from just expressions which are not tuples), but not as lvalues

**Comment by axic** [comment]
Instead of requiring every position to be defined, SOL-007 in #3326 proposes ellipsis to replace arbitrary number of arguments either as a head or tail:
```
// valid:
var ( ... , e3, e4) = (V0,V1,V2,V3,V4);
var (e0, e1, ...) = (V0,V1,V2,V3,V4);

// invalid: (?)
var (e0, ..., e4) = (V0,V1,V2,V3,V4);
```

**Comment by chriseth** [comment]
When we remove `var`, we should also remove this feature. The intention is that you want code to break if the interface of a function suddenly changes.

**Comment by jtakalai** [comment]
So... what's the resolution on how tuple assignments are done in the future? Instead of 
```
var (,a,,,,b,,,,,c,,,,,) = crazyfunction()
```
we'd have to first introduce the types like so:
```
uint a;
string b;
bytes32 c;
(,a,,,,b,,,,,c,,,,,) = crazyfunction()
```
I don't see how that is an improvement.

**Comment by axic** [comment]
This is a safety improvement.

**Comment by d10r** [comment]
I also think it should be possible to keep destructuring assignments oneliners.  
`(uint a, string b) = somefunction();`
seems a natural choice and wouldn't be less safe than the multi-line version.

For elements to be ignored, a syntax [like in Rust](https://doc.rust-lang.org/book/second-edition/ch18-03-pattern-syntax.html#ignoring-values-in-a-pattern) (`_` for a single, `..` for multiple values) could be a good option (I don't have a strong opinion about having or not having the *match all* variant).
`(_, string b) = somefunction();`
`_` would imo make it a bit more explicit than just the leading `,` that a value is being ignored.

**Comment by chriseth** [comment]
Decision: On the left hand side, empty component is fine (`_` is not required).

**Comment by chriseth** [comment]
Can be implemented via warning in pre-0.5.0
